runner/sub-agents.go

```diff
@@ -0,0 +1,188 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package runner
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"text/template"
+	"time"
+
+	"github.com/percona/pmm/api/agent"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm-agent/utils/logger"
+)
+
+type State int32
+
+const (
+	INVALID State = 0
+	RUNNING State = 1
+	STOPPED State = 2
+	EXITED  State = 3
+)
+
+type AgentParams struct {
+	AgentId uint32
+	Type    agent.Type
+	Args    []string
+	Env     []string
+	Configs map[string]string
+	Port    uint32
+}
+
+type SubAgent struct {
+	cmd    *exec.Cmd
+	log    *logger.CircularWriter
+	l      *logrus.Entry
+	state  State
+	params *AgentParams
+
+	restartChan  chan struct{}
+	restartCount int
+}
+
+type templateParams struct {
+	ListenPort uint32
+}
+
+func NewSubAgent(params *AgentParams) *SubAgent {
+	l := logrus.WithField("component", "runner").
+		WithField("agentID", params.AgentId).
+		WithField("type", params.Type)
+
+	return &SubAgent{
+		params: params,
+		log:    logger.New(10),
+		l:      l,
+		state:  INVALID,
+	}
+}
+
+func (m *SubAgent) Start(ctx context.Context) error {
+	if m.GetState() == RUNNING {
+		return fmt.Errorf("can't start the process, process is already running")
+	}
+	name := m.binary()
+	args, err := m.args()
+	if err != nil {
+		m.l.Errorln(err)
+		return err
+	}
+	cmd := exec.CommandContext(ctx, name, args...)
+	cmd.Env = m.params.Env
+	cmd.Stdout = io.MultiWriter(os.Stdout, m.log)
+	cmd.Stderr = io.MultiWriter(os.Stderr, m.log)
+
+	err = cmd.Start()
+	if err != nil {
+		m.state = EXITED
+		return err
+	}
+	m.cmd = cmd
+	m.state = RUNNING
+	m.restartChan = make(chan struct{})
+	go func() {
+		_ = m.cmd.Wait()
+		if m.state != STOPPED {
+			m.state = EXITED
+			close(m.restartChan)
+		}
+	}()
+	return nil
+}
+
+func (m *SubAgent) args() ([]string, error) {
+	params := templateParams{
+		ListenPort: m.params.Port,
+	}
+	var args []string
+	for _, arg := range m.params.Args {
+		buffer := &bytes.Buffer{}
```

Comment:
`buffer := new(bytes.Buffer)` or (even better) `var buffer bytes.Buffer`. There are no fields you can initialize in `bytes.Buffer` struct.